### PR TITLE
🐛 Machine deletion skips waiting for volumes detached for unreachable Nodes

### LINKED ISSUE #10661: MachineHealthCheck unable to remediate unreachable node with volumes attached
### What steps did you take and what happened?

* Create a CAPA cluster with at least one machine/node
* Apply a machinehealthcheck that attempts to remediate machines when nodes stop reporting status
```yaml
spec:
  maxUnhealthy: 2
  unhealthyConditions:
  - status: Unknown
    timeout: 8m0s
    type: Ready
```
* Run a pod on the cluster that mounts a persistent volume
* Stop the underlying EC2 instance in AWS
* Observe that the `DrainingSucceeded` status condition on the machine reports `status: "True"` after the `skipWaitForDelete` timeout during the drain is exceeded (https://github.com/kubernetes-sigs/cluster-api/blob/a2b7dd15155d6d7fa00145b2b726207043502e61/internal/controllers/machine/machine_controller.go#L672-L675)
* The machine is then stuck in a deleting state forever because the volume is not detached

### What did you expect to happen?

When a machinehealthcheck is attempting to remediate a machine when its underlying EC2 instance is stopped, I expect that it will successfully drain the node/replace the machine.

### Cluster API version

1.7.1

### Kubernetes version

v1.27.13+e709aa5

### Anything else you would like to add?

I believe that we can address this by setting `GracePeriodSeconds: 1` like OpenShift's machinehealthcheck controller:
* OpenShift: https://github.com/openshift/machine-api-operator/blob/dcf1387cb69f8257345b2062cff79a6aefb1f5d9/pkg/controller/machine/drain_controller.go#L164-L171
* CAPI: https://github.com/kubernetes-sigs/cluster-api/blob/a2b7dd15155d6d7fa00145b2b726207043502e61/internal/controllers/machine/machine_controller.go#L672-L675

because for unreachable nodes, deleting pods with a specified grace period will allow for successful volume detachment.

### Label(s) to be applied

/kind bug
/area machine

### PR DESCRIPTION
**Pull request by typeid**
**What this PR does / why we need it**:

When MHC are remediating an unresponsive node, CAPI drains the node and gives up waiting for pods to be deleted after the `SkipWaitForDeleteTimeoutSeconds`. If one of the pods it gave up on has an attached volume, the machine deletion gets stuck.

This PR aligns the behavior for volume detachment during deletion to what we're doing with the drains: if the node is unreachable, we skip detaching volumes.

**Which issue(s) this PR fixes** 
Fixes #10661

/area machine

### CODE REVIEW

**Review by mjlshen — COMMENTED** [review]
This is how OpenShift's handles this situation as well https://github.com/openshift/machine-api-operator/blob/dcf1387cb69f8257345b2062cff79a6aefb1f5d9/pkg/controller/machine/drain_controller.go#L164-L171. Furthermore, this allows pods to successfully be deleted, triggering CSI drivers to detach the relevant volumes, which allows the machine to be replaced

/lgtm

**Comment by jackfrancis** [comment]
@willie-yao @Jont828 Could you sanity check the equivalent functionality for MachinePoolMachines? Do we need changes there as well?

**Comment by sbueringer** [comment]
> @willie-yao @Jont828 Could you sanity check the equivalent functionality for MachinePoolMachines? Do we need changes there as well?

Do MachineHealthChecks work with MachinePool Machines? (I just don't remember that we implemented anything for it, at least in core CAPI)

**Comment by enxebre** [comment]
@typeid can you clarify where exactly in this https://github.com/kubernetes/kubectl/blob/master/pkg/drain/drain.go#L268-L360 flow this makes the difference?

**Comment by mjlshen** [comment]
@enxebre `GracePeriodSeconds` (the difference) is in https://github.com/kubernetes/kubectl/blob/d3ad75324522280fa8bccd7ad949b34336f5fc84/pkg/drain/drain.go#L312
--> https://github.com/kubernetes/kubectl/blob/d3ad75324522280fa8bccd7ad949b34336f5fc84/pkg/drain/drain.go#L150
--> https://github.com/kubernetes/kubectl/blob/d3ad75324522280fa8bccd7ad949b34336f5fc84/pkg/drain/drain.go#L133-L136

Skipping a couple steps here, but `GracePeriodSeconds` then takes effect in the kube-apiserver https://github.com/kubernetes/apiserver/blob/259cd1817cb1cc73bc33029df69934ac5c0d07ea/pkg/registry/generic/registry/store.go#L1116 where the `graceful` will eventually evaluate to `false` after the grace period has expired and the pods will be deleted by the api server instead of being stuck in a `Terminating` state forever.

That said, this is also behavior that could also be handled via https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#non-graceful-node-shutdown instead. The description exactly describes the symptoms of this bug as well.
> When a node is shutdown but not detected by kubelet's Node Shutdown Manager, the pods that are part of a [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) will be stuck in terminating status on the shutdown node and cannot move to a new running node. This is because kubelet on the shutdown node is not available to delete the pods so the StatefulSet cannot create a new pod with the same name. If there are volumes used by the pods, the VolumeAttachments will not be deleted from the original shutdown node so the volumes used by these pods cannot be attached to a new running node. As a result, the application running on the StatefulSet cannot function properly. If the original shutdown node comes up, the pods will be deleted by kubelet and new pods will be created on a different running node. If the original shutdown node does not come up, these pods will be stuck in terminating status on the shutdown node forever.

**Comment by enxebre** [comment]
>That said, this is also behavior that could also be handled via https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#non-graceful-node-shutdown instead. The description exactly describes the symptoms of this bug as well.

I guess for us to take advantage of it the instance would need to be in the process of being shut down already, so it makes sense to me to mimic it earlier, at least for now.

>Skipping a couple steps here, but GracePeriodSeconds then takes effect in the kube-apiserver https://github.com/kubernetes/apiserver/blob/259cd1817cb1cc73bc33029df69934ac5c0d07ea/pkg/registry/generic/registry/store.go#L1116 where the graceful will eventually evaluate to false after the grace period has expired and the pods will be deleted by the api server instead of being stuck in a Terminating state forever.

So does this effectively invalidates our SkipWaitForDeleteTimeoutSeconds, since all pods will match GracePeriodSeconds first?

**Comment by typeid** [comment]
I initially had the same understanding as @mjlshen's explanation, but I did more digging and I'm no longer sure just setting the grace period to 1 is actually solving the issue here. 

Reason:
- if no grace period[ or -1 (CAPI case)](https://github.com/kubernetes-sigs/cluster-api/blob/97d398103d93320d14d09bfc2947a7508e9d36bc/internal/controllers/machine/machine_controller.go#L654) is specified, we use the default `deleteOptions`: [see here](https://github.com/kubernetes/kubectl/blob/d3ad75324522280fa8bccd7ad949b34336f5fc84/pkg/drain/drain.go#L131-L141)
- we call the eviction api with these empty `deleteOptions`: [see here](https://github.com/kubernetes/kubectl/blob/d3ad75324522280fa8bccd7ad949b34336f5fc84/pkg/drain/drain.go#L162)
- this results in using the default grace period for the object (30 seconds for pods if unspecified):
  - `If this value is nil, the default grace period for the specified type will be used.`  [source](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.30/#eviction-v1-policy)
  - For pods, this `Defaults to 30 seconds.` , see `terminationGracePeriodSeconds` [here](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#pod-v1-core)

So I would expect the following to happen (current logic as well as with 1 second grace period - nothing differs in terms of pod deletion):
- we [evict all pods via goroutines at the same time](https://github.com/kubernetes/kubectl/blob/d3ad75324522280fa8bccd7ad949b34336f5fc84/pkg/drain/drain.go#L279-L280C49)
- we then wait for these pods to be deleted, but [timeout after 20 seconds due to our global timeout](https://github.com/kubernetes/kubectl/blob/d3ad75324522280fa8bccd7ad949b34336f5fc84/pkg/drain/drain.go#L446)
- we keep on retrying until we reach the SkipWaitForDeleteTimeoutSeconds time [we set here](https://github.com/kubernetes-sigs/cluster-api/blob/97d398103d93320d14d09bfc2947a7508e9d36bc/internal/controllers/machine/machine_controller.go#L675), then we ignore the pods not yet deleted and move on. This means we show the drain as completed, the pods are still around.

However, at this point, the eviction API has already been called (both cases, current vs PR), and the deletion grace period should be set on the pod, so the api server _should_ take care of it.
There seems to be no difference when it comes to actual pod deletion between explicitly setting it and letting it default, it should only decide the sooner or later.

I re-tested draining with grace period 1, and it doesn't delete the pods for my reproducer. I thought I had tried that, but by not working I think that confirms the logic above. Setting the grace period to 0, however, results in the pods deleting. 

Further digging in apiserver...
- With grace period 0 we run different logic [on the apiserver side](https://github.com/kubernetes/apiserver/blob/4aef12dc73bd08c68036e1f8990c869806cbab58/pkg/registry/rest/delete.go#L120), return (graceful = false, gracefulPending =false , err = nil).
- With grace period 1 (or -1, which results in deleteGracePeriod = pod setting, default 30 sec), we update the deletion timestamp and grace period: https://github.com/kubernetes/apiserver/blob/4aef12dc73bd08c68036e1f8990c869806cbab58/pkg/registry/rest/delete.go#L134  and return (graceful=true, gracefulpending=false, err = nil)

The logic after that gets a bit complex, but it seems like setting it to `0` could be what want for immediate deletion. 

I tried reproducing this with MAPI MHC, and it skips the pod deletion but doesn't get stuck replacing the node: I'm not sure what happens with the volume. Maybe it doesn't block the node deletion due to existing volumes? I'll have to dig further on that.

> So does this effectively invalidates our SkipWaitForDeleteTimeoutSeconds, since all pods will match GracePeriodSeconds and get terminated first?

Not necessarily. I think in the case the pods still don't delete even with grace period = 1, we would still skip them after their `deletionTimestamp` is older than 5 minutes with that setting. MAPI actually sets `SkipWaitForDeleteTimeoutSeconds` to 1 as well, though ([here](https://github.com/openshift/machine-api-operator/blob/dcf1387cb69f8257345b2062cff79a6aefb1f5d9/pkg/controller/machine/controller.go#L73) and [here](https://github.com/openshift/machine-api-operator/blob/dcf1387cb69f8257345b2062cff79a6aefb1f5d9/pkg/controller/machine/drain_controller.go#L169)). We might want to mimic that as well? 

I'll sync up with @mjlshen to get the same understanding and see if I'm possibly missing something here. :)

**Comment by enxebre** [comment]
>With grace period 0 we run different logic [on the apiserver side](https://github.com/kubernetes/apiserver/blob/4aef12dc73bd08c68036e1f8990c869806cbab58/pkg/registry/rest/delete.go#L120), return (graceful = false, gracefulPending =false , err = nil).
With grace period 1 (or -1, which results in deleteGracePeriod = pod setting, default 30 sec), we update the deletion timestamp and grace period: https://github.com/kubernetes/apiserver/blob/4aef12dc73bd08c68036e1f8990c869806cbab58/pkg/registry/rest/delete.go#L134 and return (graceful=true, gracefulpending=false, err = nil)

With a pending source code verification I assume this is because the second case requires kubelet coordination via pod status, which is not possible because the kubelet is unreachable.

>I tried reproducing this with MAPI MHC, and it skips the pod deletion but doesn't get stuck replacing the node: I'm not sure what happens with the volume. Maybe it doesn't block the node deletion due to existing volumes? I'll have to dig further on that.

Yes, the volume detaching decoupling form draining it's not implemented in mapi.

So to summarise, the erratic behaviour is that we have a semantic to ack an unreachable Node and an opinion on how react to it (Assume that give up on those pods after 5min is safe as eviction request is satisfied by pdb) but we are not consistent on that opinion for volumes dettachment.
So my preference would that we add that semantic consistently to the volume detachment process. I.e I propose we add additional logic within isNodeVolumeDetachingAllowed / nodeVolumeDetachTimeoutExceeded to account for unreachable Nodes and give up after the same SkipWaitForDeleteTimeoutSeconds we have for drain.

FWIW see related old discussion when we introduced this semantic https://github.com/kubernetes-sigs/cluster-api/pull/2165#discussion_r373525424

**Comment by typeid** [comment]
That sounds good to me. 

I think we should still make the drain consistent to what MAPI does for unreachable nodes: setting GracePeriod and the PodDeletionTimeout both to 1 second. While this still would not fix issue #10662 (to my current understanding), it would at least allow faster replacement of unreachable nodes. I will open a new issue for that and link a new PR to not cause extra confusion by remapping this PR.

**Comment by typeid** [comment]
@enxebre I updated the PR. 

The current `drainNode` behavior is that we mark the drain as completed even if we skipped removing pods. The check for reachability is within the drain, so after `isNodeDrainAllowed`. 
I opted for keeping this behavior for the volume detachment: if the node is unreachable, we skip removing the volumes but mark the `VolumeDetachSucceededCondition` condition as `true`. 

Does does make sense to you, or should we actually move the logic to `isNodeVolumeDetachingAllowed`?

File: internal/controllers/machine/machine_controller_test.go [hunk 1]
```diff
@@ -1599,6 +1599,117 @@ func TestIsNodeVolumeDetachingAllowed(t *testing.T) {
...
+				Status: corev1.NodeStatus{
+					Conditions: []corev1.NodeCondition{
+						{
+							Type:   corev1.NodeReady,
+							Status: corev1.ConditionUnknown,
+						},
+					},
+				},
+			},
+			expected: true,
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_test.go:1688]: false?

File: internal/controllers/machine/machine_controller_test.go [hunk 2]
```diff
@@ -1599,6 +1599,117 @@ func TestIsNodeVolumeDetachingAllowed(t *testing.T) {
...
+					Conditions: []corev1.NodeCondition{
+						{
+							Type:   corev1.NodeReady,
+							Status: corev1.ConditionUnknown,
+						},
+					},
+					VolumesAttached: attachedVolumes,
+				},
+			},
+			expected: true,
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_test.go:1671]: false?

File: internal/controllers/machine/machine_controller.go [hunk 3]
```diff
@@ -712,6 +713,14 @@ func (r *Reconciler) shouldWaitForNodeVolumes(ctx context.Context, cluster *clus
 		return true, err
 	}
 
+	if noderefutil.IsNodeUnreachable(node) {
+		// If a node is unreachable, we can't detach the volume.
+		// We need to skip the detachment as we otherwise block deletions
+		// of unreachable nodes when a volume is attached.
+		log.Info("Skipping volume detachment as node is unreachable.")
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller.go:720]: I think you meant false her?

Comment (typeid) [reply internal/controllers/machine/machine_controller.go:720]: Woops, clear coffee deficiency. Indeed, got the logic the wrong way around. We should not wait for the detachment as the comment mentions.

File: internal/controllers/machine/machine_controller_test.go [hunk 4]
```diff
@@ -1599,6 +1599,117 @@ func TestIsNodeVolumeDetachingAllowed(t *testing.T) {
...
+				},
+			},
+			expected: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			var objs []client.Object
```
Comment (enxebre) [inline internal/controllers/machine/machine_controller_test.go:1695]: you might need to rebase to pick https://github.com/kubernetes-sigs/cluster-api/pull/10692

Comment (typeid) [reply internal/controllers/machine/machine_controller_test.go:1695]: Thanks, I was really confused what's going on with that.

**Comment by enxebre** [comment]
For history context this where we introduced permanent wait on detachment:
https://github.com/kubernetes-sigs/cluster-api/issues/4707

This is where we make it API input:
https://github.com/kubernetes-sigs/cluster-api/issues/6285.

Now this PR aligns the behaviour with drain for unreachable Nodes.
/lgtm

**Comment by enxebre** [comment]
ptal @chrischdi @sbueringer

**Comment by sbueringer** [comment]
I'll take a look, but I need some time to get a full picture

**Comment by chrischdi** [comment]
Implementation makes sense to me and matches the issue

/lgtm

**Comment by sbueringer** [comment]
Just a quick update, I'm looking into this. Just have to double check something with folks downstream. I'll get back to you ASAP.

**Comment by sbueringer** [comment]
/lgtm

/hold
I want to check with some other folks, will get back to you ASAP

**Comment by sbueringer** [comment]
Thx folks!
